store: add migrate code and tests.

sgotti opened the pull request:
This patch adds migrate codes and tests. It simply updates the current db
version to 1 without changing the schema.

The migration needs an exclusive store lock to avoid concurrent old rkt versions
operating to an updated db and returning strange errors or giving bad behaviours.
As the store lock is removed when spawning the stage0, migration isn't blocked by
running containers but only by rkt processes doing store operations (like
downloading an image, rendering an aci etc...).

By now no reports are done to the user if waiting for a store global lock, this to avoid the store writing to stdout. If this is needed, probably the store functions should be changed to add a migrate bool to NewStore (or an additional CheckMigrate functions) to call so the various rkt commands can know and log that a migration is in progress.

Now migration covers only db updates. If the migration needs to also operate on
the other "non transactional" store entities (like the diskv stores or the tree
store) this should be thoroughly reasoned, as a rollback in case of an error it's not an
easy task.

Test are added to introduce the migration testing framework and be the base for
future migration tests.

Future updated will probably be introduced by #319.
I also created a test commit sgotti/rocket@9d63aff2b4ebe2915185502e8518d8b17f080045 (it adds a lastused fields to aciinfo) to show how the next version migrations and tests should be done. During these tests I noticed a couple of ql bugs that I reported in these 2 PRs (cznic/ql#92, cznic/ql#93).

jonboulle commented on store/migrate.go:
```diff
@@ -0,0 +1,37 @@
+package store
```
Please add the copyright

sgotti replied:
Whoops. Noticed that also store/tree*.go and store/schema.go have missing copyright headers. I'll open another PR for them.

jonboulle commented on store/migrate.go:
```diff
@@ -0,0 +1,37 @@
+
+	for v := version + 1; v <= finalVersion; v++ {
+		err := migrateTable[v](tx)
+		if err != nil {
```
if err == nil { 
  err = updateDBVersion(tx, v)
}

sgotti replied:
Good. Done!

jonboulle commented on store/migrate_test.go:
```diff
@@ -0,0 +1,382 @@
+package store
```
Copyright header

jonboulle commented on store/migrate_test.go:
```diff
@@ -0,0 +1,382 @@
+// The ACIInfo struct for different db versions. The ending VX_Y represent the
+// first and the last version where the format isn't changed
+// The latest existing struct should to be updated when updating the db version
+// without chaning the struct format (ex. V0_1 to V0_2).
```
changing

jonboulle commented on store/migrate_test.go:
```diff
@@ -0,0 +1,382 @@
+// The latest existing struct should to be updated when updating the db version
+// without chaning the struct format (ex. V0_1 to V0_2).
+// A new struct and its relative function should be added if the format is changed.
+// The same applies for all the the other structs.
```
all of

jonboulle commented on store/store.go:
```diff
@@ -114,27 +131,48 @@ func NewStore(base string) (*Store, error) {
-		}
-		if version > dbVersion {
-			return fmt.Errorf("Current cas db version: %d greater than the current rkt expected version: %d", version, dbVersion)
+		} else {
```
could you shortcircuit return instead of else, please?

jonboulle commented on store/store.go:
```diff
@@ -114,27 +131,48 @@ func NewStore(base string) (*Store, error) {
+		if err != nil {
+			return nil, err
+		}
+		// TODO(sgotti) take a db backup (for debugging and last resort rollback?)
```
Yeah, this would be prudent (although we can do in follow up if not part of this review)

sgotti replied:
ok!

jonboulle replied:
https://github.com/coreos/rkt/issues/748

jonboulle commented on store/store.go:
```diff
@@ -114,27 +131,48 @@ func NewStore(base string) (*Store, error) {
+		}
+		// TODO(sgotti) take a db backup (for debugging and last resort rollback?)
+		fn := func(tx *sql.Tx) error {
+			err := migrate(tx, dbVersion)
```
return migrate(tx, dbVersion)

jonboulle commented on store/db.go:
```diff
@@ -104,6 +104,10 @@ func (db *DB) DoTx(fns ...txfunc) error {
 		return err
 	}
 	defer func() {
+		if r := recover(); r != nil {
```
hm, what is this for exactly?

sgotti replied:
I noticed that, if a called function panics, the deferred function was going to call a tx.Commit (as on panic err is nil), with the possibility to commit an incomplete transaction, instead of doing rollbacking or returning an error (that will rollback on the next db.Close).

Thinking more about it, I reworked it to just not use a deferred function (as there's only one error path).

jonboulle commented:
@sgotti this is really great! Really appreciate the thoroughness and care, looks like great groundwork for future migrations.

sgotti commented:
@jonboulle Thanks for your review! I addressed your comments and added two additional checks in migrate.go to check that the migration function for version V is defined in the migrate map and that the reuested finalVersion isn't greater than actual dbVersion.

jonboulle commented:
:+1:, thanks so much!
